File: functional/faster/faster.jl
function slow(x)
    "slow-$(x)"
end

heavy (x) = "heavy-$x"

function faster(f::Function, x, cache) 
    if haskey(cache, x) 
        a = cache[x]
        return "from cache $a"    
    else 
        a = f(x) 
        cache[x] = a
        return "placed in cache $a"   
    end 
end

function makeFunc(f::Function) 
    cache = Dict()
    x-> faster(f, x, cache)
end

println(slow(4))
println(heavy(4))

f1 = makeFunc(slow)
f2 = makeFunc(heavy)

println(f1(55))
println(f2(55))
println(f1(55))
println(f2(55))
println(f1(55))
println(f2(55))

File: functional/faster/faster.scala
/* write a faster function that caches slower function results */
val slow = (a:Int) => { 
    val b = a +2
    "slow-" + b.toString 
}

def heavy (a:Int) : String  = {
    return "heavy-" + a.toString
}

def makeFaster(f:Int => String): (Int) => String = {
    val cache = collection.mutable.Map[Int, String]()
    return (x: Int) => {
        var a = ""
        if (cache.contains(x))  {
            a = "from cache : " + cache(x)
        } else {
            a = f(x)
            cache(x) = a
            a = "write cache : " + a 
        }
        a
    }  
}

println("...")
println(slow(23))
println(heavy(23))

val f1 = makeFaster(slow)
val f2 = makeFaster(heavy)

println("...")
println(f1(24))
println(f1(24))
println(f2(24))
println(f2(24))
println(f2(24))

